R/git-default-branch.R

```diff
@@ -0,0 +1,408 @@
+  repo <- git_repo()
+  maybe_string(old_name)
+  check_string(new_name)
+
+  # if (!is.null(old_name) &&
+  #     !gert::git_branch_exists(old_name, local = TRUE, repo = repo)) {
+  #   ui_stop("Can't find existing local branch named {ui_value(old_name)}")
+  # }
+
+  cfg <- github_remote_config(github_get = TRUE)
+  check_for_config(cfg, ok_configs = c("ours", "fork", "no_github"))
+  ui_info("GitHub remote configuration type: {ui_value(cfg$type)}")
+  ui_info("
+    Read more about GitHub remote configurations at:
+    {ui_value('https://happygitwithr.com/common-remote-setups.html')}")
+
+  # TODO: handle no_github case and exit
+  # cfg is now either fork or ours
+
+  tr <- target_repo(cfg, role = "source", ask = FALSE)
+  old_source_db <- tr$default_branch
+  ui_info("
+    Source repo is {ui_value(tr$repo_spec)} and its current default branch is \\
+    {ui_value(old_source_db)}.")
+
+  if (!isTRUE(tr$can_admin)) {
+    ui_stop("
+      You don't seem to have {ui_field('admin')} permissions for \\
+      {ui_value(tr$repo_spec)}, which is required to rename the default \\
+      branch.")
+  }
+
+  old_name <- old_name %||% guess_local_default_branch()
+
+  if (old_name != old_source_db) {
+    ui_oops("
+      It's weird that the current default branch for your local repo and \\
+      the source repo are different:
+      {ui_value(old_name)} (local) != {ui_value(old_source_db)} (source)")
+    if (ui_nope(
+      "Are you sure you want to proceed?",
+      yes = "yes", no = "no", shuffle = FALSE)) {
+      ui_stop("Cancelling.")
+    }
+  }
+
+  if (new_name == old_source_db) {
+    ui_info("
+      Source repo {ui_value(tr$repo_spec)} already has \\
+      {ui_value(old_source_db)} as its default branch.")
+  } else {
+    gh <- gh_tr(tr)
+    ui_done("
+      Renaming {ui_value(old_source_db)} branch to {ui_value(new_name)} in \\
+      the source repo {ui_value(tr$repo_spec)}.")
+    gh(
+      "POST /repos/{owner}/{repo}/branches/{from}/rename",
+      from = old_source_db,
+      new_name = new_name
+    )
+  }
+
+  rediscover_default_branch(old_name = old_name, verbose = FALSE)
+}
+
+git_default_branch_remote <- function(remote = "origin") {
+  url <- git_remotes()[[remote]]
+  if (length(url) == 0) {
+    ui_stop("There is no remote named {ui_value(remote)}.")
+  }
+
+  # TODO: generalize here for GHE hosts that don't include 'github'
+  parsed <- parse_github_remotes(url)
+  if (grepl("github", parsed$host)) {
+    return(github_remotes(remote, github_get = TRUE)$default_branch)
+  }
+
+  repo <- git_repo()
+  res <- tryCatch(
+    {
+      gert::git_fetch(remote = remote, repo = repo, verbose = FALSE)
+      gert::git_remote_ls(remote = remote, verbose = FALSE, repo = repo)
+    },
+    error = function(e) NA_character_
+  )
+  if (is.data.frame(res)) {
+    res <- path_file(res$symref[res$ref == "HEAD"])
+  }
+  res
+}
+
+guess_local_default_branch <- function(ask = is_interactive()) {
```

Comment:
This is what (maybe) needs to become more integrated with the exported `git_default_branch()`.